src/query/src/window_sort.rs

```diff
@@ -0,0 +1,3101 @@
+    DisplayAs, DisplayFormatType, ExecutionPlan, ExecutionPlanProperties, PlanProperties,
+};
+use datafusion_common::utils::bisect;
+use datafusion_common::{internal_err, DataFusionError};
+use datafusion_physical_expr::{Partitioning, PhysicalSortExpr};
+use datatypes::value::Value;
+use futures::Stream;
+use itertools::Itertools;
+use snafu::ResultExt;
+use store_api::region_engine::PartitionRange;
+
+use crate::error::{QueryExecutionSnafu, Result};
+
+/// A complex stream sort execution plan which accepts a list of `PartitionRange` and
+/// merge sort them whenever possible, and emit the sorted result as soon as possible.
+/// This sorting plan only accept sort by ts and will not sort by other fields.
+///
+/// internally, it call [`streaming_merge`] multiple times to merge multiple sorted "working ranges"
+///
+/// the input stream must be concated in the order of `PartitionRange` in `ranges`(and each `PartitionRange` is sorted within itself), or else the result will be incorrect
+#[derive(Debug, Clone)]
+pub struct WindowedSortExec {
+    /// Physical sort expressions(that is, sort by timestamp)
+    expression: PhysicalSortExpr,
+    /// Optional number of rows to fetch. Stops producing rows after this fetch
+    fetch: Option<usize>,
+    /// The input ranges indicate input stream will be composed of those ranges in given order
+    ranges: Vec<PartitionRange>,
+    /// All available working ranges and their corresponding working set
+    ///
+    /// working ranges promise once input stream get a value out of current range, future values will never be in this range
+    all_avail_working_range: Vec<(TimeRange, BTreeSet<usize>)>,
+    input: Arc<dyn ExecutionPlan>,
+    /// Execution metrics
+    metrics: ExecutionPlanMetricsSet,
+    properties: PlanProperties,
+}
+
+fn check_partition_range_monotonicity(ranges: &[PartitionRange], descending: bool) -> bool {
+    if descending {
+        ranges.windows(2).all(|w| w[0].end >= w[1].end)
+    } else {
+        ranges.windows(2).all(|w| w[0].start <= w[1].start)
+    }
+}
+
+impl WindowedSortExec {
+    pub fn try_new(
+        expression: PhysicalSortExpr,
+        fetch: Option<usize>,
+        ranges: Vec<PartitionRange>,
+        input: Arc<dyn ExecutionPlan>,
+    ) -> Result<Self> {
+        if !check_partition_range_monotonicity(&ranges, expression.options.descending) {
+            let msg = if expression.options.descending {
+                "Input `PartitionRange`s's upper bound is not monotonic non-increase"
+            } else {
+                "Input `PartitionRange`s's lower bound is not monotonic non-decrease"
+            };
+            let plain_error = PlainError::new(msg.to_string(), StatusCode::Unexpected);
+            return Err(BoxedError::new(plain_error)).context(QueryExecutionSnafu {});
+        }
+
+        let properties = PlanProperties::new(
+            input.equivalence_properties().clone(),
+            Partitioning::UnknownPartitioning(1),
```

Comment:
This plan won't collapse partitions.

```suggestion
            input.output_partitioning().clone(),
```